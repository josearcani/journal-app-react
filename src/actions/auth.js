import { getAuth, signInWithPopup } from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';

// async action
export const startLoginEmailPassword = (email, password) => {
  // use fetch or axios to get data

  return (dispatch) => {
    setTimeout(() => {
      dispatch( login('asfasi23534on32', 'Gerardo') )
    }, 1500);
  }
}

export const startGoogleLogin = () => {
  return (dispatch) => {
    const auth = getAuth();
    signInWithPopup(auth, googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName))
      })
  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
})
