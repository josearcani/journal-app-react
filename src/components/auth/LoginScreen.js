import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';

import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';

import validator from 'validator';
import { removeError, setError } from '../../actions/ui';

import { useForm } from '../../hooks/useForm';


export const LoginScreen = () => {
  
  const dispatch = useDispatch();

  const { msgError, loading } = useSelector(state => state.ui);

  const [ formValues, handleInputChange ] = useForm({
    email: 'test1@test.com',
    password: 'secret',
  })
  
  const { email, password } = formValues;

  const handleSubmit = (e) => {
    e.preventDefault();

    if (isFormValid()) {
      dispatch(startLoginEmailPassword(email, password))
    }
  }

  const handleGoogleLogin = () => {
    dispatch( startGoogleLogin() );
  }

  const isFormValid = () => {
    if (!validator.isEmail(email)) {
      dispatch(setError('Email is not valid'))
      return false
    } else if (password.trim().length === 0) {
      dispatch(setError('no valid password'))
      return false
    }
    dispatch(removeError())
    return true
  }

  return (
    <>
      <h3 className="auth__title">Login</h3>
      <form
        onSubmit={ handleSubmit }
        className="animate__animated animate__fadeIn"
      >
        { msgError && <div className="auth__alert-error">
          { msgError }
        </div> }
        <input
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="auth__input"
          value={ email }
          onChange={ handleInputChange }
        />

        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          value={ password }
          onChange={ handleInputChange }
        />

        <button
          type="submit"
          className="btn btn-primary btn-block"
          disabled={ loading }
        >
          Login
        </button>

        <div className="auth__social-networks">
          <p>Login with social networks</p>

          <div 
            className="google-btn"
            onClick={ handleGoogleLogin }
          >
            <div className="google-icon-wrapper">
              <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
            </div>
            <p className="btn-text">
              Sign in with google
            </p>
          </div>
        </div>
        <Link to="/auth/register" className="link">
          Create new account
        </Link>
      </form>
    </>
  )
}
