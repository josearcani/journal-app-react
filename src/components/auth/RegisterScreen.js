import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';

export const RegisterScreen = () => {

  const dispatch = useDispatch();

  // the cb inside useSelector returns the entire state { auth, ui }
  const { msgError } = useSelector( state => state.ui);

  // console.log( msgError);

  const [ formValues, handleInputChange ] = useForm({
    name: 'Joseph',
    email: 'test1@test.com',
    password: 'secret',
    password2: 'secret',
  })

  const { name, email, password, password2 } = formValues;

  const handleRegister = (e) => {
    e.preventDefault();

    if (isFormValid()) {
      dispatch(startRegisterWithEmailPasswordName(email, password, name));
    }

  }

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch( setError('name is required') )
      return false

    } else if (!validator.isEmail(email)) {
      dispatch( setError('Email is not valid') )
      return false

    } else if (!validator.isLength(password, { min: 5, max: 15 })) {
      dispatch( setError('password should be at least 6 characters') )
      return false

    } else if (password !== password2) {
      dispatch( setError('password should be the same') )
      return false

    }
    // true only when is valid
    dispatch( removeError())
    return true
  }

  return (
    <>
    <h3 className="auth__title">Register</h3>
    <form onSubmit={ handleRegister }>

      { msgError && <div className="auth__alert-error">
        { msgError }
      </div> }

      <input
        type="text"
        placeholder="Name"
        name="name"
        autoComplete="off"
        className="auth__input"
        value={ name }
        onChange={ handleInputChange }
      />

      <input
        type="email"
        placeholder="Email"
        name="email"
        autoComplete="off"
        className="auth__input"
        value={ email }
        onChange={ handleInputChange }
      />

      <input
        type="password"
        placeholder="Password"
        name="password"
        className="auth__input"
        value={ password }
        onChange={ handleInputChange }
      />

      <input
        type="password"
        placeholder="Confirm password"
        name="password2"
        autoComplete="off"
        className="auth__input"
        value={ password2 }
        onChange={ handleInputChange }
      />

      <button
        type="submit"
        className="btn btn-primary btn-block mb-1"
        // disabled={true}
      >
        Register
      </button>

      
      <Link to="/auth/login" className="link">
        Already registered?
      </Link>
    </form>
  </>
  )
}
