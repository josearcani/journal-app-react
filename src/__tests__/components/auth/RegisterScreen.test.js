import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';

import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
  auth: {},
  ui: {
    loading: false,
    msgError: null
  }
};

let store = mockStore(initState);

const wrapper = mount(
  <Provider store={ store }>
    <MemoryRouter>
      <RegisterScreen />
    </MemoryRouter>
  </Provider>
);

describe('Test on <RegisterScreen /> component', () => {
  
  beforeEach(() => {
    store.clearActions()
    // store = mockStore(initState);
    // jest.clearAllMocks();
  });
  
  test('should RegisterScreen render normally', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should submit fail due to empty fields', () => {
    wrapper.find('input[type="email"]').simulate('change', {
      target: { name: 'email', value: '' }
    });
    // console.log(wrapper.find('input[type="email"]').html());
    wrapper.find('form').simulate('submit', {
      preventDefault(){}
    })

    const actions = store.getActions();

    expect(actions[0]).toEqual({
      type: types.uiSetError,
      payload: 'Email is not valid'
    });
  });

  test('should show alert error message', () => {
    const initState = {
      auth: {},
      ui: {
        loading: false,
        msgError: 'Email not correct!!'
      }
    };
    
    const store = mockStore(initState);

    const wrapper = mount(
      <Provider store={ store }>
        <MemoryRouter>
          <RegisterScreen />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
    expect(wrapper.find('.auth__alert-error').text().trim()).toBe('Email not correct!!');
  });
});
