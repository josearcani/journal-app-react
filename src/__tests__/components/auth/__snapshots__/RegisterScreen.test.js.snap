// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <RegisterScreen /> component should RegisterScreen render normally 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Joseph"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value="test1@test.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="secret"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="secret"
    />
    <button
      className="btn btn-primary btn-block mb-1"
      disabled={false}
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
